lib/proxy/ftp/conn.c

```diff
@@ -134,16 +134,27 @@ conn_t *proxy_ftp_conn_accept(pool *p, conn_t *data_conn, conn_t *ctrl_conn,
 conn_t *proxy_ftp_conn_connect(pool *p, const pr_netaddr_t *bind_addr,
     const pr_netaddr_t *remote_addr, int frontend_data) {
   conn_t *conn, *opened = NULL;
-  int res, reverse_dns;
+  int default_inet_family = 0, remote_family, res, reverse_dns, xerrno;
 
   if (p == NULL ||
       remote_addr == NULL) {
     errno = EINVAL;
     return NULL;
   }
 
+  remote_family = pr_netaddr_get_family(remote_addr);
+  pr_trace_msg(trace_channel, 9,
+    "using %s family for backend socket address %s",
+    remote_family == AF_INET ? "IPv4" : "IPv6",
+    pr_netaddr_get_ipstr(remote_addr));
+  default_inet_family = pr_inet_set_default_family(p, remote_family);
```

Comment:
This is the actual fix for the issue: we make sure to tell the core Inet API, explicitly, the address family to use for the socket opened for our backend passive data transfer.  When `UseIPv6 on` is in effect, we _would_ be using IPv6 for this, even if the backend data address (obtained via `PASV` response) is an explicit IPv4 address.  Such a mismatch would cause `connect(2)` to return `EINVAL`, even for seemingly-valid addresses, as reported in this issue.